### PR TITLE
Move volume helpers to "k8s.io/component-helpers/storage/volume".

### PR DESCRIPTION
**Pull request by ikeeip**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This patch aims to simplify decoupling `pkg/scheduler/framework/plugins` package from internal `k8s.io/kubernetes` packages. 

#### Which issue(s) this PR fixes:

Refs: #89930, #102953

#### Special notes for your reviewer:

Some helpers from `k8s.io/kubernetes/pkg/controller/volume/persistentvolume` package moved to `k8s.io/component-helpers/storage/volume` package:

- IsDelayBindingMode
- GetBindVolumeToClaim
- IsVolumeBoundToClaim
- FindMatchingVolume
- CheckVolumeModeMismatches
- CheckAccessModes
- GetVolumeNodeAffinity

Also "CheckNodeAffinity" from `k8s.io/kubernetes/pkg/volume/util` package moved to `k8s.io/component-helpers/storage/volume` package to prevent diamond dependency conflict.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go [hunk 1]
```diff
@@ -16,7 +16,14 @@ limitations under the License.
...
 package volume
 
-import v1 "k8s.io/api/core/v1"
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/component-helpers/scheduling/corev1"
+	"k8s.io/klog/v2"
```
Comment (liggitt) [inline staging/src/k8s.io/component-helpers/storage/volume/helpers.go:25]: logging inside a helper function is surprising... shouldn't the caller decide whether to log or not?

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go [hunk 2]
```diff
@@ -16,7 +16,14 @@ limitations under the License.
 
 package volume
 
-import v1 "k8s.io/api/core/v1"
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/component-helpers/scheduling/corev1"
```
Comment (liggitt) [inline staging/src/k8s.io/component-helpers/storage/volume/helpers.go:24]: cross-package use of helpers is unexpected... should this function just live in `k8s.io/component-helpers/scheduling/corev1`?

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 1 above)
Comment (msau42) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:25]: This particular log line is at level 10. I am ok removing it.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go [hunk 3]
```diff
@@ -225,19 +222,17 @@ func FindMatchingVolume(
 		}
 
 		// check if PV's DeletionTimeStamp is set, if so, skip this volume.
-		if utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection) {
-			if volume.ObjectMeta.DeletionTimestamp != nil {
-				continue
-			}
+		if enableStorageObjectInUseProtection && volume.ObjectMeta.DeletionTimestamp != nil {
```
Comment (msau42) [inline staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:225]: Should be cleaned up in a separate PR, but this feature has been GA since 1.11 and we can remove the feature gate.

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 2 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:24]: fine with me to have it in scheduling

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 1 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:25]: +1

File: pkg/controller/volume/persistentvolume/binder_test.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ import (
 	v1 "k8s.io/api/core/v1"
 	storage "k8s.io/api/storage/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
+	pvutil "k8s.io/component-helpers/storage/volume"
```
Comment (alculquicondor) [inline pkg/controller/volume/persistentvolume/binder_test.go:25]: remove the alias

File: pkg/controller/volume/persistentvolume/index_test.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	ref "k8s.io/client-go/tools/reference"
 	featuregatetesting "k8s.io/component-base/featuregate/testing"
-	pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
+	pvutil "k8s.io/component-helpers/storage/volume"
```
Comment (alculquicondor) [inline pkg/controller/volume/persistentvolume/index_test.go:30]: remove alias if possible

File: pkg/controller/volume/persistentvolume/index_test.go [hunk 6]
```diff
@@ -1524,8 +1524,10 @@ func TestFindMatchVolumeWithNode(t *testing.T) {
 		},
 	}
 
+	enableStorageObjectInUseProtection := utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection)
+
 	for name, scenario := range scenarios {
-		volume, err := pvutil.FindMatchingVolume(scenario.claim, volumes, scenario.node, scenario.excludedVolumes, true)
+		volume, err := pvutil.FindMatchingVolume(scenario.claim, volumes, scenario.node, scenario.excludedVolumes, true, enableStorageObjectInUseProtection)
```
Comment (alculquicondor) [inline pkg/controller/volume/persistentvolume/index_test.go:1530]: Shouldn't this test be in the `k8s.io/component-helpers/storage/volume` package?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 7]
```diff
@@ -40,11 +40,10 @@ import (
 	"k8s.io/client-go/util/workqueue"
 	cloudprovider "k8s.io/cloud-provider"
 	volerr "k8s.io/cloud-provider/volume/errors"
-	storagehelpers "k8s.io/component-helpers/storage/volume"
+	pvutil "k8s.io/component-helpers/storage/volume"
```
Comment (alculquicondor) [inline pkg/controller/volume/persistentvolume/pv_controller.go:43]: remove alias. Otherwise, `storagehelpers` was better.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 8]
```diff
@@ -40,11 +40,11 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/workqueue"
 	cloudprovider "k8s.io/cloud-provider"
+	pvutil "k8s.io/component-helpers/storage/volume"
```
Comment (alculquicondor) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go:43]: remove alias

File: staging/src/k8s.io/component-helpers/storage/volume/helpers_test.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+			t.Errorf("CheckTopology %v returned error: %v", c.name, err)
+		}
+		if err == nil && !c.expectSuccess {
+			t.Errorf("CheckTopology %v returned success, expected error", c.name)
+		}
+	}
+}
+
+func testVolumeWithNodeAffinity(t *testing.T, affinity *v1.VolumeNodeAffinity) *v1.PersistentVolume {
+	objMeta := metav1.ObjectMeta{Name: "test-constraints"}
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/helpers_test.go:210]: nit: inline this in line 212

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 3 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:225]: oh, let's do that first

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go [hunk 10]
```diff
@@ -225,19 +222,17 @@ func FindMatchingVolume(
...
 
 		nodeAffinityValid := true
 		if node != nil {
 			// Scheduler path, check that the PV NodeAffinity
 			// is satisfied by the node
-			// volumeutil.CheckNodeAffinity is the most expensive call in this loop.
+			// CheckNodeAffinity is the most expensive call in this loop.
 			// We should check cheaper conditions first or consider optimizing this function.
-			err := volumeutil.CheckNodeAffinity(volume, node.Labels)
+			err := CheckNodeAffinity(volume, node.Labels)
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:235]: @liggitt if this function moves to `scheduling`, then this package would still depend on it.

I think the dependency is reasonable.

**Comment by alculquicondor** [comment]
Before we proceed:

It looks like the only user of `pkg/controller/volume/scheduling` is the scheduler.

Any reason not to move it inside `pkg/scheduler/framework/plugins/volumebinding`? I suppose it's due to ownership, but that can be fixed. @msau42 do you agree?

**Comment by msau42** [comment]
> It looks like the only user of `pkg/controller/volume/scheduling` is the scheduler.
> 
> Any reason not to move it inside `pkg/scheduler/framework/plugins/volumebinding`? 

I think that's a great idea. I don't remember the reason why we put it here originally but the current scheduler plugin directory structure looks reasonable to me.

**Comment by ikeeip** [comment]
>> It looks like the only user of pkg/controller/volume/scheduling is the scheduler.
>>Any reason not to move it inside pkg/scheduler/framework/plugins/volumebinding?
>
> I think that's a great idea. I don't remember the reason why we put it here originally but the current scheduler plugin directory structure looks reasonable to me.

@alculquicondor @msau42 I have prepared PR https://github.com/kubernetes/kubernetes/pull/103072 with this changes already

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 2 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:24]: If we do that, `pkg/volume/util/operationexecutor` will depends on `k8s.io/component-helpers/scheduling/corev1`. Is it ok?

**Comment by alculquicondor** [comment]
/hold

Let's park this until #103072 is approved

**Review by alculquicondor — COMMENTED** [review]
overall approach lgtm. Please attend the open comments.

File: pkg/controller/volume/persistentvolume/index.go [hunk 11]
```diff
@@ -86,13 +88,14 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *v1.PersistentVol
 	// example above).
 	allPossibleModes := pvIndex.allPossibleMatchingAccessModes(claim.Spec.AccessModes)
 
+	enableStorageObjectInUseProtection := utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection)
```
[the remark this answers is not in the document]
Comment (ikeeip) [reply pkg/controller/volume/persistentvolume/index.go]: I have done it in separate commit (https://github.com/kubernetes/kubernetes/pull/103062/commits/4f99101c509a9b3cc3232f3c57f2be7394b5be0f). But this feature gate is still used in:

```
$ find . -name '*.go' | xargs grep 'features.StorageObjectInUseProtection'
./k8s.io/kubernetes/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go:     c.storageObjectInUseProtection = featureGates.Enabled(features.StorageObjectInUseProtection)
./k8s.io/kubernetes/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:    if utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection) {
./k8s.io/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go:    if utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection) {
./k8s.io/kubernetes/pkg/controller/volume/persistentvolume/index_test.go:                       defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageObjectInUseProtection, true)()
./k8s.io/kubernetes/pkg/controller/volume/persistentvolume/index_test.go:                       defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageObjectInUseProtection, true)()
./k8s.io/kubernetes/cmd/kube-controller-manager/app/core.go:            utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection),
./k8s.io/kubernetes/cmd/kube-controller-manager/app/core.go:            utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection),
./k8s.io/kubernetes/cmd/kube-controller-manager/app/core.go:            utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection),
./sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go:        c.storageObjectInUseProtection = featureGates.Enabled(features.StorageObjectInUseProtection)
./sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util/util.go:           if utilfeature.DefaultFeatureGate.Enabled(features.StorageObjectInUseProtection) {
```

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/volume/persistentvolume/index.go]: Can you do it in a separate PR?

[the remark this answers is not in the document]
Comment (ikeeip) [reply pkg/controller/volume/persistentvolume/index.go]: Sure. Here it is #104903

**Comment by ikeeip** [comment]
@alculquicondor All related PR's was merged. Can we process this one?

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go [hunk 12]
```diff
@@ -42,3 +48,26 @@ func GetPersistentVolumeClass(volume *v1.PersistentVolume) string {
...
 	return volume.Spec.StorageClassName
 }
+
+// CheckNodeAffinity looks at the PV node affinity, and checks if the node has the same corresponding labels
+// This ensures that we don't mount a volume that doesn't belong to this node
+func CheckNodeAffinity(pv *v1.PersistentVolume, nodeLabels map[string]string) error {
+	return checkVolumeNodeAffinity(pv, &v1.Node{ObjectMeta: metav1.ObjectMeta{Labels: nodeLabels}})
+}
+
+func checkVolumeNodeAffinity(pv *v1.PersistentVolume, node *v1.Node) error {
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/helpers.go:58]: nit: just have one function

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go [hunk 13]
```diff
@@ -28,8 +28,6 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	storagelisters "k8s.io/client-go/listers/storage/v1"
 	"k8s.io/client-go/tools/reference"
-	storagehelpers "k8s.io/component-helpers/storage/volume"
-	volumeutil "k8s.io/kubernetes/pkg/volume/util"
 )
 
 const (
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:33]: Can you rename all of these to have the prefix `Annotation`? Or maybe it's better as suffix? Have a look at what we do for other annotations.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go [hunk 14]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package persistentvolume
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:17]: Can you double check if all the moved functions have at least more than one binary depending on it?

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go [hunk 15]
```diff
@@ -16,7 +16,13 @@ limitations under the License.
 
 package volume
 
-import v1 "k8s.io/api/core/v1"
+import (
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/storage/volume/helpers.go:19]: There doesn't seem to be a reason to have 2 files. All the functions refer to pv/pvc.

Maybe just merge them in a separate commit (or PR). The way you have the PR now helps a lot to see that nothing really changed and it's just a move.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 14 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:17]: I have checked this out and found nothing like that. The script I have used here:

```sh
make all GOLDFLAGS="" GOGCFLAGS=all="-N -l"
find _output/bin/ -type f | xargs -n1 sh -c 'gdb -batch -ex "info functions k8s.io/component-helpers/storage/volume" $0 2>/dev/null | awk -F"[.(]" -v FN=$0 "/void/{ print FN, \$4 }"' > /tmp/pkg_func_usage.txt
awk '{ print $2 }' /tmp/pkg_func_usage.txt | sort | uniq -c
```

The results here:
```
      4 CheckAccessModes
      6 CheckNodeAffinity
      4 CheckVolumeModeMismatches
      6 checkVolumeNodeAffinity
      4 claimToClaimKey
      4 FindMatchingVolume
      4 GetBindVolumeToClaim
      8 GetPersistentVolumeClaimClass
      6 GetPersistentVolumeClass
      4 IsDelayBindingMode
      3 IsDelayBindingProvisioning
      4 IsVolumeBoundToClaim
```

But I found `GetVolumeNodeAffinity` function which is used only by unit test `TestFindMatchVolumeWithNode`. I will move it closer to the test and make private.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 13 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:33]: I found three different ways how annotations looks like:

```sh
grep -r --include "*.go" --exclude-dir="vendor" -E "\sAnnotation[[:alpha:]]{2,} = " . | wc -l
10

grep -r --include "*.go" --exclude-dir="vendor" -E "\s[[:alpha:]]{2,}AnnotationKey = " . | wc -l
26

grep -r --include "*.go" --exclude-dir="vendor" -E "\s[[:alpha:]]{2,}Annotation = " . | wc -l
45
```

@alculquicondor What do you think If I suffix it with `Annotation`? For example, `AnnBindCompleted` will be `BindCompletedAnnotation`.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 14 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:17]: Oops, I found another place where `GetVolumeNodeAffinity` func is used. This is test too but it located in another package `k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding/binder_test.go`.

Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:17]: @alculquicondor What do you think is it ok if I move this function in `k8s.io/kubernetes/staging/src/k8s.io/component-helpers/storage/volume/pv_helpers_test.go` and leave it public?

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 15 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:19]: Sure I will do it another PR as a follow up.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 14 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:17]: No. Better copy it in each of the packages that need it. It's just a factory of the API. `Get` is not a very useful prefix. Maybe `build` is a better prefix.

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 13 above)
Comment (ikeeip) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:33]: There are too many changes:
```
        modified:   pkg/controller/volume/persistentvolume/binder_test.go
        modified:   pkg/controller/volume/persistentvolume/delete_test.go
        modified:   pkg/controller/volume/persistentvolume/framework_test.go
        modified:   pkg/controller/volume/persistentvolume/provision_test.go
        modified:   pkg/controller/volume/persistentvolume/pv_controller.go
        modified:   pkg/controller/volume/persistentvolume/pv_controller_base.go
        modified:   pkg/controller/volume/persistentvolume/pv_controller_test.go
        modified:   pkg/controller/volume/persistentvolume/recycle_test.go
        modified:   pkg/scheduler/framework/plugins/volumebinding/assume_cache_test.go
        modified:   pkg/scheduler/framework/plugins/volumebinding/binder.go
        modified:   pkg/scheduler/framework/plugins/volumebinding/binder_test.go
        modified:   pkg/scheduler/framework/plugins/volumebinding/test_utils.go
        modified:   pkg/scheduler/generic_scheduler_test.go
        modified:   plugin/pkg/admission/storage/persistentvolume/label/admission.go
        modified:   plugin/pkg/admission/storage/persistentvolume/label/admission_test.go
        modified:   staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go
        modified:   test/integration/scheduler_perf/scheduler_perf_legacy_test.go
        modified:   test/integration/util/util.go
        modified:   test/integration/volume/persistent_volumes_test.go
```
@alculquicondor I think, Its's better to do this in separate PR.  What do you think?

**Comment by alculquicondor** [comment]
Code freeze is over. Hopefully we can get this merged soon. Can you rebase?

**Comment by ikeeip** [comment]
> Code freeze is over. Hopefully we can get this merged soon. Can you rebase?

Great news! Rebased.

**Comment by msau42** [comment]
/lgtm
/approve

Thanks so much for cleaning this up!

File: staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go (hunk 13 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go:33]: As long as it happens before the 1.24 code freeze, fine to do in a follow up PR.

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 15 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:19]: Don't forget about this please.

File: staging/src/k8s.io/component-helpers/storage/volume/helpers.go (hunk 2 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/storage/volume/helpers.go:24]: I think the way the dependencies are working now is good.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by ikeeip** [comment]
@alculquicondor Could you /unhold this PR, please? You hold it early in favor of #103072

**Comment by alculquicondor** [comment]
I'm sorry. I forgot there was a hold.

/hold cancel
